sdkutil/event.go

```diff
@@ -87,3 +104,172 @@ func GetString(attrs []sdk.Attribute, key string) (string, error) {
 	}
 	return "", ErrNotFound
 }
+
+// BlockID provides type to properly format print blockchain ID types.
+type BlockID struct {
+	Owner    *sdk.AccAddress
+	DSeq     *uint64
+	GSeq     *uint32
+	OSeq     *uint32
+	Provider *sdk.AccAddress
+}
+
+// NewBlockID initializes the BlockID struct, passing nil values is acceptable
+// for fields which are not used.
+func NewBlockID(owner *sdk.AccAddress, dseq *uint64, gseq *uint32, oseq *uint32, provider *sdk.AccAddress) BlockID {
+	return BlockID{
+		Owner:    owner,
+		DSeq:     dseq,
+		GSeq:     gseq,
+		OSeq:     oseq,
+		Provider: provider,
+	}
+}
+
+// String method provides the full formating of all the BlockID fields which are not nil.
+// Format: [Owner AccAddress]/[DSeq]/[GSeq]/[OSeq]/[Provider AccAddress]
+// eg: akash1vgv30hr8lel8r7270wywmak5c82es3njws3u4z/795423625/1/1/akash1pxqksfr60zc6uadht6300uavwl3s58ct0apt9w
+func (b BlockID) String() string {
+	parts := make([]string, 0)
+	if b.Owner != nil {
+		parts = append(parts, b.Owner.String())
+	}
+	if b.DSeq != nil {
+		parts = append(parts, strconv.FormatUint(*b.DSeq, 10))
+	}
+	if b.GSeq != nil {
+		parts = append(parts, strconv.FormatUint(uint64(*b.GSeq), 10))
+	}
+	if b.OSeq != nil {
+		parts = append(parts, strconv.FormatUint(uint64(*b.OSeq), 10))
+	}
+	if b.Provider != nil {
+		parts = append(parts, b.Provider.String())
+	}
+	return path.Join(parts...)
+}
+
+// ReflectBlockID accepts an ID struct and unpacks the block identifying fields.
+// Useful for passing *ID types and getting consistent output based on the consistent
+// field names.
+func ReflectBlockID(id interface{}) *BlockID {
+	b := BlockID{}
+
+	val := reflect.ValueOf(id)
+	typeOfID := val.Type()
+
+	for i := 0; i < val.NumField(); i++ {
```

Comment:
if the `i`th value is `nil` then the `i+1`th value should be `nil`.

again not sure if we should enforce here.

Reply:
I'm also not sure if the ordering of val.NumField is consistent with how the struct 'looks', so I wrote this to be agnostic of any ordering, which I think is more robust.